tilequeue/rawr.py

```diff
@@ -256,26 +263,33 @@ def __call__(self):
                 rawr_tile_generated = True
 
                 with time_block(timing, 'toi-intersect'):
-                    # because this returns a generator, the timing is wrong
-                    # unless we realize immediately
-                    coords_to_enqueue = list(self.rawr_toi_intersector(coords))
+                    coords_to_enqueue, intersect_metrics = \
+                        self.rawr_toi_intersector(coords)
                 coords_intersected = True
 
                 with time_block(timing, 'queue-write'):
-                    self.queue_writer.enqueue_batch(coords_to_enqueue)
+                    n_enqueued, n_inflight = \
+                        self.queue_writer.enqueue_batch(coords_to_enqueue)
                 coords_enqueued = True
 
                 with time_block(timing, 'queue-done'):
                     self.rawr_queue.done(msg_handle)
+                msg_done = True
```

Comment:
I found the flags somewhat confusing, e.g:

```python
try:
  do_first_thing(...)
  first_thing_done = True
  do_second_thing(...)
  second_thing_done = True
  ...
except:
  if not first_thing_done:
    ...
  elif not second_thing_done:
    ...
```

Might it be clearer written out as individual try-catch blocks, perhaps wrapping the logging to avoid repetition?

```python
try:
  do_first_thing(...)
except Exception:
  self._log_exception("first thing")
  continue

try:
  do_second_thing(...)
except Exception:
  self._log_exception("second thing")
  continue

...
def _log_exception(self, msg):
  if self.logger:
    stacktrace = format_stacktrace_one_line()
    msg = 'Error: %s' % msg
    ...
```

Another possibility might be to have a single `try` block, but modify the exceptions with an internal `with` block, something like:

```python
try:
  with describe_error("first thing"):
    do_first_thing(...)
  with describe_error("second thing"):
    do_second_thing(...)

except Exception:
  if self.logger:
    self.logger.info(e.message())

class describe_error:
  def __init__(self, msg):
    self.msg = msg

  def __enter__(self):
    pass

  def __exit__(self, type, value, traceback):
    raise Exception(self.msg, parent=value, traceback=traceback)
```

What do you think?

Reply:
Agreed that this was a little clunky. It started off with just one or two flags, but then grew into many more.

I considered the `with` block route, but I wasn't immediately sure whether the stacktrace would be preserved as expected by chaining the exceptions. We could instead just mutate some state that contained the error message, similarly to the timing block, but that seemed like it was a little bit awkward or heavy handed for what it was trying to do. I decided to just wrap each block in its own try/except like you mentioned first, and although it's repetitive, it's at least clear as to what's going on.

Updated in 1548e66.